libyul/optimiser/ConstantOptimiser.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that replaces constants by expressions that compute them.
+ */
+
+#include <libyul/optimiser/ConstantOptimiser.h>
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/Metrics.h>
+#include <libyul/AsmData.h>
+#include <libyul/AsmPrinter.h>
+#include <libyul/Utilities.h>
+#include <libyul/AsmParser.h>
+
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+using Representation = ConstantOptimiser::Representation;
+
+void ConstantOptimiser::visit(Expression& _e)
+{
+	if (_e.type() == typeid(Literal))
+	{
+		Literal const& literal = boost::get<Literal>(_e);
+		if (literal.kind != LiteralKind::Number)
+			return;
+
+		if (
+			Expression const* repr =
+				RepresentationFinder(m_evmVersion, m_meter, locationOf(_e), m_cache)
+				.tryFindRepresentation(valueOfLiteral(literal))
+		)
+			_e = ASTCopier{}.translate(*repr);
+	}
+	else
+		ASTModifier::visit(_e);
+}
+
+Expression const* RepresentationFinder::tryFindRepresentation(dev::u256 const& _value)
+{
+	if (_value < 0x10000)
+		return nullptr;
+
+	Representation const& repr = findRepresentation(_value);
+	if (repr.expression->type() == typeid(Literal))
+		return nullptr;
+	else
+		return repr.expression.get();
+}
+
+Representation const& RepresentationFinder::findRepresentation(dev::u256 const& _value)
+{
+	if (m_cache.count(_value))
+		return m_cache.at(_value);
+
+	Representation routine = represent(_value);
+
+	if (dev::bytesRequired(~_value) < dev::bytesRequired(_value))
+		// Negated is shorter to represent
+		routine = min(move(routine), represent("not", findRepresentation(~_value)));
+
+	// Decompose value into a * 2**k + b where abs(b) << 2**k
+	for (unsigned bits = 255; bits > 8 && m_maxSteps > 0; --bits)
+	{
+		unsigned gapDetector = unsigned((_value >> (bits - 8)) & 0x1ff);
+		if (gapDetector != 0xff && gapDetector != 0x100)
+			continue;
+
+		u256 powerOfTwo = u256(1) << bits;
+		u256 upperPart = _value >> bits;
+		bigint lowerPart = _value & (powerOfTwo - 1);
+		if ((powerOfTwo - lowerPart) < lowerPart)
+		{
+			lowerPart = lowerPart - powerOfTwo; // make it negative
+			upperPart++;
+		}
+		if (upperPart == 0)
+			continue;
+		if (abs(lowerPart) >= (powerOfTwo >> 8))
+			continue;
+		Representation newRoutine;
```

Comment:
If we want some higher assurance here, we could start with a ``u256 checkValue = upperPart;`` here and then depending on the path modify it with ``checkValue = shlWorkaround(checkValue, bits)`` or ``checkValue *= 2 <however boost raises to power> bits`` and later ``checkValue += u256(abs(lowerPart))`` or ``checkValue -= u256(abs(lowerPart)))`` and in the end assert that ``checkValue == _value``...

But maybe it's unnecessary.

Reply:
The old constant optimiser has a mini-interpreter that runs at the end...